### PR TITLE
interop-testing: add soak test cases to interop client

### PR DESCRIPTION
**Pull request by apolcyn**
This is a Go analogue of Java PR: https://github.com/grpc/grpc-java/pull/8407

Documentation PR: https://github.com/grpc/grpc/pull/27030 adds cross language documentation for this test

FWIW, C++  [channel_soak test case is here](https://github.com/grpc/grpc/blob/3e34e3aac790a256943b4b7dada269209c60fe44/test/cpp/interop/client.cc#L55) and [rpc_soak test case is here](https://github.com/grpc/grpc/blob/3e34e3aac790a256943b4b7dada269209c60fe44/test/cpp/interop/client.cc#L72) (for which the code is mainly in https://github.com/grpc/grpc/blob/3e34e3aac790a256943b4b7dada269209c60fe44/test/cpp/interop/interop_client.cc#L1120).

RELEASE NOTES: N/A

### CODE REVIEW

File: interop/test_utils.go [hunk 1]
```diff
@@ -673,6 +676,89 @@ func DoPickFirstUnary(tc testgrpc.TestServiceClient) {
...
+		client = testgrpc.NewTestServiceClient(conn)
+	}
+	// do a large-unary RPC
+	pl := ClientNewPayload(testpb.PayloadType_COMPRESSABLE, largeReqSize)
+	req := &testpb.SimpleRequest{
+		ResponseType: testpb.PayloadType_COMPRESSABLE,
+		ResponseSize: int32(largeRespSize),
+		Payload:      pl,
+	}
+	reply, err := client.UnaryCall(context.Background(), req)
```
Comment (menghanl) [inline interop/test_utils.go:701]: I understand the idea is to not set timeouts.

Should this set deadline to be the `overallDeadline`?
We want this test to fail at the time, right?

File: interop/test_utils.go [hunk 2]
```diff
@@ -673,6 +676,89 @@ func DoPickFirstUnary(tc testgrpc.TestServiceClient) {
...
+		if err != nil {
+			totalFailures++
+			fmt.Fprintf(os.Stderr, "soak iteration: %d elapsed_ms: %d failed: %s\n", i, latencyMs, err)
+		} else if results[i].latency > perIterationMaxAcceptableLatency {
+			totalFailures++
+			fmt.Fprintf(os.Stderr, "soak iteration: %d elapsed_ms: %d exceeds max acceptable latency: %d\n", i, latencyMs, perIterationMaxAcceptableLatency.Milliseconds())
+		} else {
+			fmt.Fprintf(os.Stderr, "soak iteration: %d elapsed_ms: %d succeeded\n", i, latencyMs)
+		}
+		h.Add(latencyMs)
```
Comment (menghanl) [inline interop/test_utils.go:747]: Nit: you can reorder this, and save those `else` statements

```go
h.add(latencyMs)
if err != nil {
  totalFailures++
  fmt.Fprintf()
  continue
}

if results[i].latency > perIterationMaxAcceptableLatency {
  totalFailures++
  fmt.Fprintf()
  continue
}
fmt.Fprintf(success)
```

File: interop/test_utils.go [hunk 3]
```diff
@@ -673,6 +676,89 @@ func DoPickFirstUnary(tc testgrpc.TestServiceClient) {
...
+	}
+	var b bytes.Buffer
+	h.Print(&b)
+	fmt.Fprintln(os.Stderr, "Histogram of per-iteration latencies in milliseconds:")
+	fmt.Fprintln(os.Stderr, b.String())
+	if len(results) < soakIterations {
+		logger.Fatalf("soak test consumed all %f seconds of time and quit early, only having ran %d out of desired %d iterations. total failures: %d. max failures threshold: %d. Some or all of the iterations that did run were unexpectedly slow. See breakdown above for which iterations succeeded, failed, and why for more info.", overallDeadline.Sub(start).Seconds(), len(results), soakIterations, totalFailures, maxFailures)
+	} else if totalFailures > maxFailures {
```
Comment (menghanl) [inline interop/test_utils.go:755]: This `else if` and the next `else` are unnecessary.
Fatalf() will exit anyway.

Comment (apolcyn) [reply interop/test_utils.go]: done

File: interop/test_utils.go (hunk 2 above)
Comment (apolcyn) [reply interop/test_utils.go]: done

File: interop/test_utils.go (hunk 1 above)
Comment (apolcyn) [reply interop/test_utils.go]: good idea, done.

I figured the previous way would usually work well enough, unless the RPC got completely stuck. But your approach is ideal

**Review by menghanl — APPROVED** [review]
LGTM.

Can you rebase? We fixed some flaky tests, that should help with the github action failures.

**Comment by apolcyn** [comment]
I've rebased.

@menghanl FYI some minor changes/specifications were made to the test's spec in https://github.com/grpc/grpc/pull/27030 while this PR was open.

I can incorporate those changes into this PR, or merge this PR and then add the changes in a followup.

**Comment by menghanl** [comment]
Yeah, if you want to make more commits in this PR, that SGTM

**Comment by easwars** [comment]
@apolcyn 
Do you want to make the other changes in this PR or a new one? We can merge this one as-is.

**Comment by apolcyn** [comment]
I've updated this PR, PTAL

File: interop/test_utils.go [hunk 4]
```diff
@@ -673,6 +676,91 @@ func DoPickFirstUnary(tc testgrpc.TestServiceClient) {
 	}
 }
 
+func doOneSoakIteration(ctx context.Context, tc testgrpc.TestServiceClient, resetChannel bool, serverAddr string, dopts []grpc.DialOption) (latency time.Duration, err error) {
+	start := time.Now()
+	// per test spec, don't include channel shutdown in latency measurement
+	defer func() { latency = time.Since(start) }()
```
Comment (menghanl) [inline interop/test_utils.go:682]: If the purpose is to exclude time spent in conn.Close(), this needs to be deferred after `defer conn.Close()`

Deferred calls are executed in last-in-first-out order https://tour.golang.org/flowcontrol/13

Comment (apolcyn) [reply interop/test_utils.go]: Whoops! I meant to leverage the LIFO ordering but accidentally swapped things, thanks.

**Comment by apolcyn** [comment]
Thanks for the review @menghanl. If recent changes still look good, can you please merge? (I don't have merge access)

**Pull request merged**